File: tokenizer/plaintext/plaintext_tokenizer.cpp
#include <iostream>
#include <fstream>
#include <string>
#include <cassert>
#include <iomanip>
#include <ctype.h>
#include <cstdio>
#include "nlohmann/json.hpp"

void usage(const std::string &program) {
  std::cerr << "Usage: " << program << " [--ignore_punctuation] [--to_lower] [--ignore_numbers] [--ignore_newlines] < INPUT_FILE.txt > OUTPUT_FILE.json" << std::endl;
  exit(0);
}


int main(int argc, char* argv[]) {

  // ------------------------------
  // handle arguments
  bool ignore_punctuation = false;
  bool to_lower = false;
  bool ignore_numbers = false;
  bool ignore_newlines = false;
  for (int i = 1; i < argc; i++) {
    if (std::string(argv[i]) == std::string("--ignore_punctuation")) {
      ignore_punctuation = true;
    } else if (std::string(argv[i]) == std::string("--to_lower")) {
      to_lower = true;
    } else if (std::string(argv[i]) == std::string("--ignore_numbers")) {
      ignore_numbers = true;
    } else if (std::string(argv[i]) == std::string("--ignore_newlines")) {
      ignore_newlines = true;
    } else {
      std::cerr << "ERROR: Unknown option '" << argv[i] << "'" << std::endl;
      usage(argv[0]);
    }
  }

  // ------------------------------
  // helper variables
  nlohmann::json tokens;
  char c;
  std::string token;
  int row = 1;
  int col = 1;
  int start_row = -1;
  int start_col = -1;
  bool last_was_alpha = false;
  bool last_was_digit = false;

  // ------------------------------
  // loop to read the input file file
  while (std::cin >> std::noskipws >> c) {
    bool is_punctuation = !isspace(c) && !std::isdigit(c) && !std::isalpha(c);

    if ((unsigned int)(c) > 127) {
      // FIXME: for now, just skip utf-8 characters since nlohmann dump gets stuck
      continue;
    }

    // ------------------------------
    // decide when to break the current string
    // break on spaces, punctuation (any symbol), or if we switch between letters and numbers
    if (isspace(c) ||
        is_punctuation ||
        (last_was_alpha && std::isdigit(c)) ||
        (last_was_digit && std::isalpha(c))) {
      if (token != "") {
        // save this token!
        std::map<std::string,nlohmann::json> tmp;
        tmp["line"]=start_row;
        tmp["char"]=start_col;
        if (last_was_digit) {
          assert (!last_was_alpha);
          tmp["type"]="number";
          tmp["value"]=std::stoi(token);
        } else {
          assert (last_was_alpha);
          tmp["type"]="string";
          tmp["value"]=token;
        }
        tokens.push_back(tmp);
        token="";
        last_was_alpha = false;
        last_was_digit = false;
      }
    }

    // ------------------------------
    // decide whether to add this character to the current string
    if (isspace(c)) {
      // never add spaces
    }
    // ------------------------------
    else if (is_punctuation) {
      assert (token == "");
      assert (last_was_alpha == false);
      assert (last_was_digit == false);
      // only add punctuation if its not being ignored
      // (punctuation is always a single symbol character per token)
      if (!ignore_punctuation) {
        std::map<std::string,nlohmann::json> tmp;
        tmp["line"]=row;
        tmp["char"]=col;
        tmp["type"]="punctuation";
        tmp["value"]=std::string(1,c);
        tokens.push_back(tmp);
      }
    }
    // ------------------------------
    else if (std::isdigit(c)) {
      assert (last_was_alpha == false);
      // only add digits/numbers if they are not being ignored
      // numbers will be 1 or more digits 0-9.
      // We break tokens between letters & numbers.
      if (!ignore_numbers) {
        if (token=="") {
          start_row = row;
          start_col = col;
          last_was_digit=true;
        } else {
          assert (last_was_digit == true);
        }
        token.push_back(c);
      }
    }
    // ------------------------------
    else if (isalpha(c)) {
      assert (last_was_digit == false);
      // string tokens will be 1 or more letters a-z or A-Z
      if (token=="") {
        start_row = row;
        start_col = col;
        last_was_alpha=true;
      } else {
        assert (last_was_alpha == true);
      }
      // option to lowercase
      if (to_lower) {
        c = std::tolower(c);
      }
      token.push_back(c);
    }

    // ------------------------------
    if (c == '\n') {
      // advance to the next row/line
      assert (token=="");
      if (!ignore_newlines) {
        // output a token for the newline
        std::map<std::string,nlohmann::json> tmp;
        tmp["line"]=row;
        tmp["char"]=col;
        tmp["type"]="newline";
        tmp["value"]="\n";
        tokens.push_back(tmp);
      }
      row++;
      col=1;
    } else {
      // advance to the next column/character
      col++;
    }
  } 

  // ------------------------------
  if (token != "") {
    // save the last token (if there was no space or newline at the end of the file)
    std::map<std::string,nlohmann::json> tmp;
    tmp["line"]=start_row;
    tmp["char"]=start_col;
    if (last_was_digit) {
      assert (!last_was_alpha);
      tmp["type"]="number";
      tmp["value"]=std::stoi(token);
    } else {
      assert (last_was_alpha);
      tmp["type"]="string";
      tmp["value"]=token;
    }
    tokens.push_back(tmp);
  }

  // ------------------------------
  // export/save in json format
  std::cout << tokens.dump(4) << std::endl;
}

File: tokenizer/plaintext/plaintext_tokenizer.py
import json
import argparse
import string


def parse_args():
    parser = argparse.ArgumentParser(description="Plaintext Tokenizer")
    parser.add_argument("input_file")
    parser.add_argument("--ignore_punctuation", action='store_true')
    parser.add_argument("--to_lower", action='store_true')
    parser.add_argument("--ignore_numbers", action='store_true')
    parser.add_argument("--ignore_newlines", action='store_true')
    return parser.parse_args()


def main():
    args = parse_args()

    with open(args.input_file, "r", encoding="ISO-8859-1") as input_file:
        file_data = input_file.read()

    tokens = []
    token = ""
    row = 1
    col = 1
    start_row = -1
    start_col = -1
    last_was_alpha = False
    last_was_digit = False
    for char in file_data:
        is_punctuation = char in string.punctuation

        # ----------------------------------------------------------------------
        # decide when to break the current string
        # break on spaces, punctuation (any symbol), or if we switch between letters and numbers
        if (char.isspace() or is_punctuation
                or (last_was_alpha and char.isdigit()) or (last_was_digit and char.isalpha())):
            # save this token
            if token != "":
                tmp = {}
                tmp["line"] = start_row
                tmp["char"] = start_col
                if (last_was_digit):
                    assert not last_was_alpha
                    # check to see if this will fit in an int -- otherwise treat it as a string
                    try:
                        tmp["value"] = int(token)
                        tmp["type"] = "number"
                    except Exception:
                        tmp["value"] = token
                        tmp["type"] = "string"
                else:
                    assert last_was_alpha
                    tmp["value"] = token
                    tmp["type"] = "string"

                tokens.append(tmp)
                token = ""
                last_was_alpha = False
                last_was_digit = False

        # ----------------------------------------------------------------------
        # decide whether to add this character to the current string
        if char.isspace():
            pass  # never add spaces
        # ------------------------------
        elif is_punctuation:
            assert token == ""
            assert not last_was_alpha
            assert not last_was_digit

            # only add punctuation if its not being ignored
            # (punctuation is always a single symbol character per token)
            if not args.ignore_punctuation:
                tmp = {}
                tmp["line"] = row
                tmp["char"] = col
                tmp["type"] = "punctuation"
                tmp["value"] = char
                tokens.append(tmp)
        # ------------------------------
        elif char.isdigit():
            assert not last_was_alpha
            # only add digits/numbers if they are not being ignored
            # numbers will be 1 or more digits 0-9.
            # We break tokens between letters & numbers.
            if not args.ignore_numbers:
                if token == "":
                    start_row = row
                    start_col = col
                    last_was_digit = True
                else:
                    assert last_was_digit
                token += char
        # ------------------------------
        elif char.isalpha():
            assert not last_was_digit
            if token == "":
                start_row = row
                start_col = col
                last_was_alpha = True
            else:
                assert last_was_alpha

            # to_lower argument
            if args.to_lower:
                char = char.lower()

            token += char

        # ----------------------------------------------------------------------
        if char == "\n":
            assert token == ""
            if not args.ignore_newlines:
                # output a token for the newline
                tmp = {}
                tmp["line"] = row
                tmp["char"] = col
                tmp["type"] = "newline"
                tmp["value"] = "\n"
                tokens.append(tmp)
            # advance to the next row/line
            row += 1
            col = 1
        else:
            col += 1
    # --------------------------------------------------------------------------
    # print the json file
    print(json.dumps(tokens, indent=4, sort_keys=True))


# Run tokenizer
main()
